overlord/hookstate/hookmgr.go

```diff
@@ -143,13 +153,60 @@ func (m *HookManager) doRunHook(task *state.Task, tomb *tomb.Tomb) error {
 		return err
 	}
 
-	// TODO: Actually dispatch the hook.
+	// Actually run the hook
+	_, err = RunHook(setup.Snap, setup.Revision, setup.Hook, tomb)
+	if err != nil {
+		if handlerErr := handler.Error(err); handlerErr != nil {
+			return handlerErr
+		}
 
-	// Done with the hook. TODO: Check the result, if success call Done(), if
-	// error, call Error(). Since we have no hooks, for now we just call Done().
-	if err := handler.Done(); err != nil {
+		return err
+	}
+
+	// Assuming no error occurred, notify the handler that the hook has
+	// finished.
+	if err = handler.Done(); err != nil {
 		return err
 	}
 
 	return nil
 }
+
+func doRunHook(snapName string, revision snap.Revision, hookName string, tomb *tomb.Tomb) ([]byte, error) {
+	runner, err := startCommand("snap", "run", snapName, "--hook", hookName, "-r", revision.String())
+	if err != nil {
+		return nil, err
+	}
+
+	hookCompleted := make(chan struct{})
+	var hookOutput []byte
+	var hookError error
+	go func() {
+		hookOutput, hookError = runner.Wait()
+		close(hookCompleted)
+	}()
+
+	select {
+	case <-hookCompleted:
+		return hookOutput, hookError
+	case <-tomb.Dying():
+		if err := runner.Kill(); err != nil {
+			return nil, fmt.Errorf("failed to abort hook: %s", err)
+		}
+		return nil, fmt.Errorf("hook was aborted")
+	}
+}
+
+func doStartCommand(name string, args ...string) ([]byte, error) {
+	output, err := exec.Command(name, args...).CombinedOutput()
+	if err != nil {
+		// Make the error a bit more informative
```

Comment:
This is leaking our internal organization. For the user "snap run --hook ..." makes no sense.

Let's please do something like this:
```
if err != nil {
        return fmt.Errorf("hook %q failed: %v", hookName, osutil.OutputErr(output, err))
}
```
Note it's `hookName` rather than `name`. We'll need a new parameter.

Can we please put this function inside `osutil`:
```
// OutputErr formats an error based on output if its length not zero,
// or returns err otherwise.
func OutputErr(output []byte, err error) error {
	output = bytes.TrimSpace(output)
	if len(output) > 0 {
		if bytes.Contains(output, []byte{'\n'}) {
			err = fmt.Errorf("\n-----\n%s\n-----", output)
		} else {
			err = fmt.Errorf("%s", output)
		}
	}
	return err
}
```

This comes from spread, and although simple it's been evolving over the years. Often handy.